Tools/clinic/cpp.py

```diff
@@ -62,15 +66,15 @@ def close(self):
         if self.stack:
             self.fail("Ended file while still in a preprocessor conditional block!")
 
-    def write(self, s):
+    def write(self, s: str):
         for line in s.split("\n"):
             self.writeline(line)
 
-    def writeline(self, line):
+    def writeline(self, line: str):
         self.line_number += 1
         line = line.strip()
 
-        def pop_stack():
+        def pop_stack() -> TokenAndCondition | NoReturn:
```

Comment:
Everything could be annotated with `| NoReturn` in Python -- there's always the possibility somebody will do `CTRL-C` in the middle of program execution :p

Unions with `NoReturn` are basically always redundant, I think:

```suggestion
        def pop_stack() -> TokenAndCondition:
```

Reply:
TIL!